parser/v2/templelementparser.go

```diff
@@ -0,0 +1,134 @@
+package parser
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/a-h/lexical/parse"
+)
+
+var templElementExpressionStartParser = parse.Or(parse.String("<! "), parse.String("<!"))
+var templElementExpressionEndParser = parse.Or(parse.String(" >"), parse.String(">"))
+var templElementExpressionCloseTagParser = parse.Or(parse.String(" />"), parse.String("/>"))
+
+var templBlockElementExpression templBlockElementExpressionParser
+
+type templBlockElementExpressionParser struct{}
+
+func (p templBlockElementExpressionParser) Parse(pi parse.Input) parse.Result {
+	var r TemplElementExpression
+
+	// Check the prefix first.
+	prefixResult := parse.Rune('@')(pi)
+	if !prefixResult.Success {
+		return prefixResult
+	}
+
+	// Once we have a prefix, we must have an expression that returns a template.
+	from := NewPositionFromInput(pi)
+	pr := parse.StringUntil(parse.All(parse.WithStringConcatCombiner, openBraceWithOptionalPadding))(pi)
+	if pr.Error != nil && pr.Error != io.EOF {
+		return pr
+	}
+	if !pr.Success {
+		return parse.Failure("templElementParser", newParseError("templ element: unterminated (missing closing '{\n')", from, NewPositionFromInput(pi)))
+	}
+	r.Expression = NewExpression(pr.Item.(string), from, NewPositionFromInput(pi))
+
+	from = NewPositionFromInput(pi)
+	if te := expressionEnd(pi); !te.Success {
+		return parse.Failure("templElementParser", newParseError("templ element: unterminated (missing closing '{')", from, NewPositionFromInput(pi)))
+	}
+
+	// Once we've had the start of a for block, we must conclude the block.
+
+	// Eat newline.
+	if lb := newLine(pi); lb.Error != nil {
+		return lb
+	}
+
+	// Node contents.
+	from = NewPositionFromInput(pi)
+	pr = newTemplateNodeParser(parse.Or(elseExpression.Parse, closeBraceWithOptionalPadding)).Parse(pi)
+	if pr.Error != nil && pr.Error != io.EOF {
+		return pr
+	}
+	// If there's no match, there's a problem in the template nodes.
+	if !pr.Success {
+		return parse.Failure("templElementParser", newParseError(fmt.Sprintf("@%s: expected nodes, but none were found", r.Expression.Value), from, NewPositionFromInput(pi)))
+	}
+
+	r.Children = pr.Item.([]Node)
+
+	// Read the required closing brace.
+	if ie := closeBraceWithOptionalPadding(pi); !ie.Success {
+		return parse.Failure("templElementParser", newParseError(fmt.Sprintf("@%s: missing end (expected '}')", r.Expression.Value), from, NewPositionFromInput(pi)))
+	}
+	return parse.Success("templElementParser", r, nil)
+}
+
+var templSelfClosingElementExpression templSelfClosingElementExpressionParser
+
+type templSelfClosingElementExpressionParser struct{}
```

Comment:
With this one, is there a need for it now?

I assume that `{! template() }` still works for templates that don't have children.

Is it that `@template()` and `@template() {\n}` can also be used?

Reply:
Ah, yes, I can see that in the tests.